eclair-core/src/main/scala/fr/acinq/eclair/payment/send/CompactBlindedPathsResolver.scala

```diff
@@ -0,0 +1,62 @@
+package fr.acinq.eclair.payment.send
+
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.{ActorContext, Behaviors}
+import akka.actor.{ActorRef, typed}
+import fr.acinq.bitcoin.scalacompat.Crypto.PublicKey
+import fr.acinq.eclair.crypto.Sphinx.RouteBlinding.BlindedRoute
+import fr.acinq.eclair.payment.send.CompactBlindedPathsResolver._
+import fr.acinq.eclair.payment.{PaymentBlindedContactInfo, PaymentBlindedRoute}
+import fr.acinq.eclair.router.Router
+import fr.acinq.eclair.wire.protocol.OfferTypes.{BlindedPath, CompactBlindedPath, PaymentInfo}
+
+import scala.annotation.tailrec
+
+object CompactBlindedPathsResolver {
+  sealed trait Command
+
+  case class Resolve(replyTo: typed.ActorRef[Seq[PaymentBlindedRoute]], blindedPaths: Seq[PaymentBlindedContactInfo]) extends Command
+
+  private case class WrappedNodeId(nodeId_opt: Option[PublicKey]) extends Command
+
+  def apply(router: ActorRef): Behavior[Command] = {
+    Behaviors.receivePartial {
+      case (context, Resolve(replyTo, blindedPaths)) =>
+        val resolver = new CompactBlindedPathsResolver(replyTo, router, context)
+        resolver.resolveCompactBlindedPaths(blindedPaths, Nil)
+    }
+  }
+}
+
+private class CompactBlindedPathsResolver(replyTo: typed.ActorRef[Seq[PaymentBlindedRoute]],
+                                          router: ActorRef,
+                                          context: ActorContext[Command]) {
+  @tailrec
+  private def resolveCompactBlindedPaths(toResolve: Seq[PaymentBlindedContactInfo],
+                                         resolved: Seq[PaymentBlindedRoute]): Behavior[Command] = {
+    if (toResolve.isEmpty) {
```

Comment:
nit: we can flatten that by using pattern matching on `headOption` (done in #2814).